File: server/src/main/java/com/antigravity/converters/CameraConfigConverter.java
package com.antigravity.converters;

import com.antigravity.proto.CameraInterfaceConfig;
import com.antigravity.protocols.camera.CameraConfig;
import com.antigravity.protocols.camera.LaneDetectionGate;
import java.util.ArrayList;
import java.util.List;

public class CameraConfigConverter {

  public static CameraInterfaceConfig toProto(CameraConfig config) {
    if (config == null) {
      return CameraInterfaceConfig.getDefaultInstance();
    }
    CameraInterfaceConfig.Builder builder =
        CameraInterfaceConfig.newBuilder()
            .setName(config.name != null ? config.name : "")
            .setInterfaceIndex(config.interfaceIndex)
            .setTargetFps(config.targetFps)
            .setAutoDetectLanes(config.autoDetectLanes);

    if (config.gates != null) {
      for (LaneDetectionGate gate : config.gates) {
        if (gate != null) {
          builder.addGates(
              com.antigravity.proto.LaneDetectionGate.newBuilder() // fqn-collision
                  .setLaneIndex(gate.laneIndex)
                  .setXPct(gate.xPct)
                  .setYPct(gate.yPct)
                  .setWidthPct(gate.widthPct)
                  .setHeightPct(gate.heightPct)
                  .setGateType(gate.gateType)
                  .setSensitivity(gate.sensitivity)
                  .build());
        }
      }
    }
    return builder.build();
  }

  public static CameraConfig fromProto(CameraInterfaceConfig protoConfig) {
    if (protoConfig == null) {
      return null;
    }
    CameraConfig config = new CameraConfig();
    config.name = protoConfig.getName();
    config.interfaceIndex = protoConfig.getInterfaceIndex();
    config.targetFps = protoConfig.getTargetFps() > 0 ? protoConfig.getTargetFps() : 60;
    config.autoDetectLanes = protoConfig.getAutoDetectLanes();

    List<LaneDetectionGate> gates = new ArrayList<>();
    if (protoConfig.getGatesList() != null) {
      for (com.antigravity.proto.LaneDetectionGate g : // fqn-collision
          protoConfig.getGatesList()) {
        LaneDetectionGate gate = new LaneDetectionGate();
        gate.laneIndex = g.getLaneIndex();
        gate.xPct = g.getXPct();
        gate.yPct = g.getYPct();
        gate.widthPct = g.getWidthPct();
        gate.heightPct = g.getHeightPct();
        gate.gateType = g.getGateType();
        gate.sensitivity = g.getSensitivity();
        gates.add(gate);
      }
    }
    config.gates = gates;
    return config;
  }
}

File: client/src/app/converters/camera_config.converter.ts
import { CameraConfig } from "@app/models/camera_config";
import { ICameraInterfaceConfig } from "@app/proto/antigravity";

export class CameraConfigConverter {
  static fromProto(proto: ICameraInterfaceConfig): CameraConfig {
    return {
      name: proto.name || "Camera Interface",
      interfaceIndex: proto.interfaceIndex ?? 0,
      targetFps: proto.targetFps || 60,
      autoDetectLanes: proto.autoDetectLanes ?? false,
      gates: (proto.gates || []).map((g) => ({
        laneIndex: g.laneIndex ?? 0,
        xPct: g.xPct ?? 0,
        yPct: g.yPct ?? 0,
        widthPct: g.widthPct ?? 0,
        heightPct: g.heightPct ?? 0,
        gateType: g.gateType ?? 0,
        sensitivity: g.sensitivity ?? 0.5,
      })),
      connectionType: (proto.connectionType as "local" | "remote") || "local",
    };
  }

  static toProto(config: CameraConfig): ICameraInterfaceConfig {
    return {
      name: config.name,
      interfaceIndex: config.interfaceIndex,
      targetFps: config.targetFps,
      autoDetectLanes: config.autoDetectLanes,
      gates: (config.gates || []).map((g) => ({
        laneIndex: g.laneIndex,
        xPct: g.xPct,
        yPct: g.yPct,
        widthPct: g.widthPct,
        heightPct: g.heightPct,
        gateType: g.gateType,
        sensitivity: g.sensitivity,
      })),
      connectionType: config.connectionType || "local",
    };
  }
}
